iroha/src/torii.rs

```diff
@@ -178,35 +225,53 @@ async fn handle_instructions(
 async fn handle_queries(
     state: State<ToriiState>,
     _path_params: PathParams,
-    _query_params: QueryParams,
+    query_params: QueryParams,
     request: HttpRequest,
 ) -> Result<HttpResponse, String> {
     //TODO: Remove when `Result::flatten` https://github.com/rust-lang/rust/issues/70142 will be stabilized
-    match SignedQueryRequest::try_from(request.body) {
+    let range = match Pagination::try_from(&query_params) {
+        Ok(range) => range,
+        Err(e) => {
+            log::error!("Failed to decode pagination: {}", e);
+            // Is it really internal server error?
+            return Ok(HttpResponse::internal_server_error());
+        }
+    };
+
+    let request = match SignedQueryRequest::try_from(request.body) {
         //TODO: check query permissions based on signature?
-        Ok(request) => match VerifiedQueryRequest::try_from(request) {
-            Ok(request) => {
-                match request
-                    .query
-                    .execute(&*state.read().await.world_state_view.read().await)
-                {
-                    Ok(value) => {
-                        let result = &QueryResult(value);
-                        Ok(HttpResponse::ok(Headers::new(), result.into()))
-                    }
-                    Err(e) => {
-                        log::error!("Failed to execute query: {}", e);
-                        Ok(HttpResponse::internal_server_error())
-                    }
-                }
-            }
-            Err(e) => {
-                log::error!("Failed to verify Query Request: {}", e);
-                Ok(HttpResponse::internal_server_error())
-            }
-        },
+        Ok(request) => request,
         Err(e) => {
             log::error!("Failed to decode transaction: {}", e);
+            return Ok(HttpResponse::internal_server_error());
+        }
+    };
+    let request = match VerifiedQueryRequest::try_from(request) {
+        Ok(request) => request,
+        Err(e) => {
+            log::error!("Failed to verify Query Request: {}", e);
+            return Ok(HttpResponse::internal_server_error());
+        }
+    };
+
+    match request
+        .query
+        .execute(&*state.read().await.world_state_view.read().await)
+    {
+        Ok(Value::Vec(value)) => {
+            let result = &QueryResult(Value::Vec(match range {
+                Pagination::From(range) => value[range].to_vec(),
+                Pagination::FromTo(range) => value[range].to_vec(),
+                Pagination::Full => value,
+            }));
+            Ok(HttpResponse::ok(Headers::new(), result.into()))
+        }
```

Comment:
For pagination it is very important to enforce order. We need to check the implementations of query execution and for the ones returning Vec sort them by IDs.